File: ls/service/src/main/java/org/opencds/cqf/cql/ls/service/LanguageClientAppender.java
package org.opencds.cqf.cql.ls.service;

import java.util.Objects;
import org.eclipse.lsp4j.MessageParams;
import org.eclipse.lsp4j.MessageType;
import org.eclipse.lsp4j.services.LanguageClient;
import ch.qos.logback.classic.Level;
import ch.qos.logback.classic.spi.ILoggingEvent;
import ch.qos.logback.core.AppenderBase;

/**
 * This class allows logging to the LSP client's {@code} logMessage} API
 */
public class LanguageClientAppender extends AppenderBase<ILoggingEvent> {
    private final LanguageClient client;

    /**
     * Constructs a LanguageClientAppender
     *
     * @param client the client to log to
     */
    public LanguageClientAppender(LanguageClient client) {
        this.client = client;
    }

    @Override
    protected void append(ILoggingEvent eventObject) {
        if (eventObject == null) {
            return;
        }

        this.client.logMessage(createMessageParams(eventObject));
    }

    MessageParams createMessageParams(ILoggingEvent eventObject) {
        Objects.requireNonNull(eventObject);

        return new MessageParams(toType(eventObject.getLevel()), eventObject.getMessage());
    }

    MessageType toType(Level level) {
        if (level == Level.ERROR) {
            return MessageType.Error;
        } else if (level == Level.WARN) {
            return MessageType.Warning;
        } else if (level == Level.INFO) {
            return MessageType.Info;
        } else {
            return MessageType.Log;
        }
    }
}

File: ls/server/src/main/kotlin/org/opencds/cqf/cql/ls/service/LanguageClientAppender.kt
package org.opencds.cqf.cql.ls.service

import ch.qos.logback.classic.Level
import ch.qos.logback.classic.spi.ILoggingEvent
import ch.qos.logback.core.AppenderBase
import org.eclipse.lsp4j.MessageParams
import org.eclipse.lsp4j.MessageType
import org.eclipse.lsp4j.services.LanguageClient

class LanguageClientAppender(private val client: LanguageClient) : AppenderBase<ILoggingEvent>() {
    override fun append(eventObject: ILoggingEvent?) {
        if (eventObject == null) return
        client.logMessage(createMessageParams(eventObject))
    }

    internal fun createMessageParams(eventObject: ILoggingEvent): MessageParams =
        MessageParams(toType(eventObject.level), eventObject.formattedMessage)

    internal fun toType(level: Level): MessageType =
        when (level) {
            Level.ERROR -> MessageType.Error
            Level.WARN -> MessageType.Warning
            Level.INFO -> MessageType.Info
            else -> MessageType.Log
        }
}
